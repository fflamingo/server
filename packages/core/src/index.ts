export function route(path: string) {
  return path;
}
