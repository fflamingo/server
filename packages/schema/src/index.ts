export function makeSchema(str: string) {
  return str;
}
