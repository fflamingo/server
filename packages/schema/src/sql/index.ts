import * as astBuilder from './sqlAstBuilder';
import * as astTypes from './sqlAstTypes';
export { findAstFromAliasName } from './sqlAstTraversal';
export { sqlAstCompile as astCompile } from './sqlAstCompile';

export { astBuilder };
export { astTypes };
