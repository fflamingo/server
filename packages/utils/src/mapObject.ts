export function mapObject<T, U, K extends string>(
  obj: Record<K, T>,
  transform: (key: string, value: T) => U
): Record<K, U> {
  return (Object.keys(obj) as [K]).reduce(
    (mapped, key) => {
      mapped[key] = transform(key, obj[key]);
      return mapped;
    },
    {} as Record<K, U>
  );
}
