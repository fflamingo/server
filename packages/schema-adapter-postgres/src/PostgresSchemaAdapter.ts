import { Schema, Adapter, sqlAstCompile } from '@fflamingo/schema';
import { GraphQLResolveInfo } from 'graphql';
import { graphqlToQuery } from './query-ast/graphqlToQuery';
import * as Knex from 'knex';
import util from 'util';

export class PostgresSchemaAdapter extends Adapter {
  database: Knex;

  constructor(database: Knex) {
    super();
    this.database = database;
  }

  async fromAST(info: GraphQLResolveInfo) {
    const ast = await graphqlToQuery(info);
    console.log('ast is', util.inspect(ast, false, null));
    const query = sqlAstCompile(this.database, ast) as Knex.QueryBuilder;
    console.log('result from query is\n', query.toString());
    const result = await query;
    console.log('query result is', util.inspect(result, false, null));
    return result.result;
  }
}
