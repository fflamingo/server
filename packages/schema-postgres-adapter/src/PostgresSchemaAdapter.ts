import { Schema, Adapter, sqlAstCompile } from '@fflamingo/schema';
import { GraphQLResolveInfo } from 'graphql';
import { graphqlToQuery } from './query-ast/graphqlToQuery';
import knex from 'knex';
import util from 'util';

export class PostgresSchemaAdapter extends Adapter {
  database;
  constructor() {
    super();
    this.database = knex({ client: 'pg' });
  }

  async fromAST(info: GraphQLResolveInfo) {
    const ast = await graphqlToQuery(info);
    console.log('ast is', util.inspect(ast, false, null));
    const query = sqlAstCompile(this.database, ast);
    console.log('result from query is\n', query.toSQL());
    return [{ email: 'aa aa', name: 'BBB' }];
  }
}
