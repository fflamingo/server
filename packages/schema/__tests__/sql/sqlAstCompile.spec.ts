import knex from 'knex';
import { sqlAstBuilder as sql, sqlAstCompile } from '../../src';

const db = knex({ client: 'pg' });

test('should create a select expression', () => {
  expect(
    sqlAstCompile(
      db,
      sql.astSelect(sql.astTable('users'), [
        sql.astField(sql.astIdentifier('email')),
        sql.astField(sql.astIdentifier('name'))
      ])
    ).toString()
  ).toMatchSnapshot();
});

test('should create a wrapped query expression', () => {
  expect(
    sqlAstCompile(
      db,
      sql.astSelect(
        sql.astWrappedQuery(
          sql.astSelect(sql.astTable('users'), [
            sql.astField(sql.astIdentifier('original_name'))
          ]),
          'original_users'
        ),
        [
          sql.astField(
            sql.astIdentifier('original_users', 'name'),
            'original_name'
          ),
          sql.astField('original_name')
        ]
      )
    ).toString()
  ).toMatchSnapshot();
});

test('should compile an aggregate function', () => {
  expect(
    sqlAstCompile(
      db,
      sql.astAggregateField('row_to_json', 'column', 'final')
    ).toString()
  ).toEqual(`row_to_json("column") as "final"`);
});
