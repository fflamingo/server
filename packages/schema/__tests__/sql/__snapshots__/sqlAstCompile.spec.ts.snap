// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a select expression 1`] = `"select \\"email\\", \\"name\\" from \\"users\\""`;

exports[`should create a wrapped query expression 1`] = `"select \\"original_users\\".\\"name\\" as \\"original_name\\", \\"original_name\\" from (select \\"original_name\\" from \\"users\\") as \\"original_users\\""`;
